ddtrace/_trace/processor/__init__.py

```diff
@@ -324,6 +335,14 @@ def on_span_finish(self, span):
 
                 if should_partial_flush:
                     log.debug("Partially flushing %d spans for trace %d", num_finished, span.trace_id)
+                    if not finished:
+                        log_msg = f"No spans to flush in partial flush for span: {span}"
+                        # Avoid potential crashes if, for some reason, we have no spans to flush even though
+                        # should_partial_flush is True.
+                        if config._telemetry_enabled:
+                            telemetry.telemetry_writer.add_log("WARNING", log_msg)
+                        log.warning(log_msg)
+                        return
```

Comment:
I worry about the state of `num_finished` here, since we incremented it by one at the start.


```python
if num_finished != trace.num_finished:
    log_msg = f"Trying to flush fewer spans than expected {num_finished} != {trace.num_finished} when finishing spans {span}"
    ...

# We are holding a lock, and we removed all the finished spans from `trace`, so there shouldn't be any finished spans left
trace.num_finished = 0

# We checked and logged already
if not finished:
    return

if should_partial_flush:
    finished[0].set_metric("_dd.py.partial_flush", num_finished)
```

What do you think about this general approach/control flow change?